File: backend/recipe_parser.py
import requests
import extruct
from bs4 import BeautifulSoup
from w3lib.html import get_base_url
import json

class RecipeParser:
    def fetch_recipe(self, url):
        """Fetches the URL and returns the HTML content."""
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        try:
            response = requests.get(url, headers=headers, timeout=10)
            response.raise_for_status()
            return response.text
        except requests.RequestException as e:
            print(f"Error fetching URL: {e}")
            return None

    def extract_recipe_title(self, html, url):
        """Extracts recipe title from Schema.org JSON-LD or HTML."""
        # Try JSON-LD first
        base_url = get_base_url(html, url)
        try:
            data = extruct.extract(html, base_url=base_url, syntaxes=['json-ld'])
            json_ld_data = data.get('json-ld', [])
            
            for item in json_ld_data:
                if item.get('@type') == 'Recipe':
                    return item.get('name', '')
                
                if '@graph' in item:
                    for node in item['@graph']:
                        if node.get('@type') == 'Recipe':
                            return node.get('name', '')
        except Exception:
            pass
        
        # Fallback to HTML title
        soup = BeautifulSoup(html, 'html.parser')
        title_tag = soup.find('title')
        if title_tag:
            return title_tag.get_text(strip=True)
        
        return "Recipe"

    def extract_ingredients_json_ld(self, html, url):
        """Extracts ingredients from Schema.org JSON-LD."""
        base_url = get_base_url(html, url)
        try:
            data = extruct.extract(html, base_url=base_url, syntaxes=['json-ld'])
        except Exception as e:
            print(f"Error in extruct: {e}")
            return []
        
        ingredients = []
        
        json_ld_data = data.get('json-ld', [])
        
        for item in json_ld_data:
            # Check for direct Recipe type
            if item.get('@type') == 'Recipe':
                ingredients = item.get('recipeIngredient', [])
                if ingredients:
                    return ingredients
            
            # Check for graph
            if '@graph' in item:
                for node in item['@graph']:
                    if node.get('@type') == 'Recipe':
                        ingredients = node.get('recipeIngredient', [])
                        if ingredients:
                            return ingredients
                            
        return ingredients

    def extract_ingredients_html(self, html):
        """Fallback extraction from HTML."""
        soup = BeautifulSoup(html, 'html.parser')
        ingredients = []
        
        # Common patterns for ingredients
        potential_lists = soup.find_all(['ul', 'ol'])
        for lst in potential_lists:
            classes = lst.get('class', [])
            id_attr = lst.get('id', '')
            
            if any('ingredient' in str(cls).lower() for cls in classes) or 'ingredient' in str(id_attr).lower():
                for li in lst.find_all('li'):
                    text = li.get_text(separator=' ', strip=True)
                    text = ' '.join(text.split())
                    ingredients.append(text)
                if ingredients:
                    return ingredients

        # Try finding container divs
        potential_divs = soup.find_all('div')
        for div in potential_divs:
            classes = div.get('class', [])
            id_attr = div.get('id', '')
            
            if any('ingredient' in str(cls).lower() for cls in classes) or 'ingredient' in str(id_attr).lower():
                items = div.find_all('li')
                if items:
                    for item in items:
                        text = item.get_text(separator=' ', strip=True)
                        text = ' '.join(text.split())
                        ingredients.append(text)
                    if ingredients:
                        return ingredients
                
        return ingredients

    def parse(self, url):
        """Main function to parse recipe from URL."""
        html = self.fetch_recipe(url)
        if not html:
            return {'ingredients': [], 'title': ''}
        
        print("Extracting recipe title...")
        title = self.extract_recipe_title(html, url)
        
        print("Attempting JSON-LD extraction...")
        ingredients = self.extract_ingredients_json_ld(html, url)
        
        if not ingredients:
            print("JSON-LD failed or not found. Attempting HTML fallback...")
            ingredients = self.extract_ingredients_html(html)
            
        return {
            'ingredients': ingredients,
            'title': title
        }

if __name__ == "__main__":
    # Test with a sample URL
    test_url = input("Enter recipe URL: ")
    if test_url:
        parser = RecipeParser()
        results = parser.parse(test_url)
        print(f"\nRecipe: {results.get('title', 'Unknown')}")
        print("\nFound Ingredients:")
        for ing in results.get('ingredients', []):
            print(f"- {ing}")

File: extension/recipe_parser.js
/**
 * Recipe Parser - JavaScript version
 * Extracts recipe ingredients and title from HTML content
 */

class RecipeParser {
    // No constructor needed

    /**
     * Extracts recipe title from JSON-LD or HTML
     */
    extractRecipeTitle(html) {
        // Try JSON-LD first
        const jsonLdScripts = html.match(/<script[^>]*type=["']application\/ld\+json["'][^>]*>([\s\S]*?)<\/script>/gi);

        if (jsonLdScripts) {
            for (const script of jsonLdScripts) {
                try {
                    const jsonText = script.replace(/<script[^>]*>/, '').replace(/<\/script>/, '');
                    const data = JSON.parse(jsonText);

                    // Handle single object
                    if (data['@type'] === 'Recipe' && data.name) {
                        return data.name;
                    }

                    // Handle @graph array
                    if (data['@graph']) {
                        for (const node of data['@graph']) {
                            if (node['@type'] === 'Recipe' && node.name) {
                                return node.name;
                            }
                        }
                    }
                } catch (e) {
                    // Continue to next script
                }
            }
        }

        // Fallback to HTML title
        const titleMatch = html.match(/<title[^>]*>(.*?)<\/title>/i);
        if (titleMatch) {
            return this.decodeHtml(titleMatch[1].trim());
        }

        return 'Recipe';
    }

    /**
     * Extracts ingredients from JSON-LD
     */
    extractIngredientsJsonLd(html) {
        const jsonLdScripts = html.match(/<script[^>]*type=["']application\/ld\+json["'][^>]*>([\s\S]*?)<\/script>/gi);

        if (!jsonLdScripts) {
            return [];
        }

        for (const script of jsonLdScripts) {
            try {
                const jsonText = script.replace(/<script[^>]*>/, '').replace(/<\/script>/, '');
                const data = JSON.parse(jsonText);

                // Handle single Recipe object
                if (data['@type'] === 'Recipe' && data.recipeIngredient) {
                    return data.recipeIngredient;
                }

                // Handle @graph array
                if (data['@graph']) {
                    for (const node of data['@graph']) {
                        if (node['@type'] === 'Recipe' && node.recipeIngredient) {
                            return node.recipeIngredient;
                        }
                    }
                }
            } catch (e) {
                // Continue to next script
            }
        }

        return [];
    }

    /**
     * Fallback: Extract ingredients from HTML
     */
    extractIngredientsHtml(html) {
        const parser = new DOMParser();
        const doc = parser.parseFromString(html, 'text/html');
        const ingredients = [];

        // Find lists with "ingredient" in class or id
        const lists = doc.querySelectorAll('ul, ol');
        for (const list of lists) {
            const classes = list.className.toLowerCase();
            const id = (list.id || '').toLowerCase();

            if (classes.includes('ingredient') || id.includes('ingredient')) {
                const items = list.querySelectorAll('li');
                for (const item of items) {
                    const text = item.textContent.trim().replace(/\s+/g, ' ');
                    if (text) {
                        ingredients.push(text);
                    }
                }
                if (ingredients.length > 0) {
                    return ingredients;
                }
            }
        }

        // Try finding divs with ingredient classes
        const divs = doc.querySelectorAll('div');
        for (const div of divs) {
            const classes = div.className.toLowerCase();
            const id = (div.id || '').toLowerCase();

            if (classes.includes('ingredient') || id.includes('ingredient')) {
                const items = div.querySelectorAll('li');
                if (items.length > 0) {
                    for (const item of items) {
                        const text = item.textContent.trim().replace(/\s+/g, ' ');
                        if (text) {
                            ingredients.push(text);
                        }
                    }
                    if (ingredients.length > 0) {
                        return ingredients;
                    }
                }
            }
        }

        return ingredients;
    }

    /**
     * Decode HTML entities
     */
    decodeHtml(html) {
        const txt = document.createElement('textarea');
        txt.innerHTML = html;
        return txt.value;
    }

    /**
     * Main parse function
     * @param {string} html - The HTML content to parse
     */
    async parse(html) {
        if (!html) {
            return { ingredients: [], title: '' };
        }

        console.log('[RecipeParser] Extracting recipe title...');
        const title = this.extractRecipeTitle(html);

        console.log('[RecipeParser] Attempting JSON-LD extraction...');
        let ingredients = this.extractIngredientsJsonLd(html);

        if (!ingredients || ingredients.length === 0) {
            console.log('[RecipeParser] JSON-LD failed. Attempting HTML fallback...');
            ingredients = this.extractIngredientsHtml(html);
        }

        return {
            ingredients: ingredients,
            title: title
        };
    }
}

// Export for use in panel.js
window.RecipeParser = RecipeParser;
